### PR TITLE
Read the default encoding configuration from the tokenizer

### PR DESCRIPTION
**Pull request by nkasmanoff**
As discussed in https://github.com/Lightning-AI/lit-gpt/issues/545, there's a potential enhancement for these base models if we make sure to use the `BOS` token for fine-tuning & generation if the base model also uses the `BOS` token at the start of it's prompts.

I'm still not 100% sure which models require this, but the way I'm saving it down is in shown in the diff, where it's now an additional argument to the model config, where the default behavior is `False`, as it was previously. 

So far I've made the change to Llama2 and a few of it's variants, where I could clearly see `BOS` was supposed to be added see (https://huggingface.co/meta-llama/Llama-2-7b-hf/blob/main/tokenizer_config.json#L2) 

While no comprehensive testing was done, I tested generation on Llama2 7B with and without the BOS token prepended to the prompt, and it was noticeable how much more on topic the output was, so I do think this will help!

Made no changes to the `EOS` token's behavior, as I don't think it's necessary. 

Thought this approach was the best way to keep this model specific and default to the original behavior, but open to adjusting process.

### CODE REVIEW

File: lit_gpt/tokenizer.py
Comment (carmocca) [inline lit_gpt/tokenizer.py:64]: You could instead make this `Optional[bool]` so that passing a boolean is interpreted as overriding the default

File: lit_gpt/config.py
Comment (carmocca) [inline lit_gpt/config.py:27]: I would prefer that we read this off the `tokenizer_config` instead of adding this setting to the model.

File: lit_gpt/tokenizer.py
Comment (carmocca) [inline lit_gpt/tokenizer.py:1]: Some tokenizers will need to read this value: https://huggingface.co/EleutherAI/pythia-70m/blob/main/tokenizer_config.json#L2 whereas others https://huggingface.co/meta-llama/Llama-2-7b-hf/blob/main/tokenizer_config.json#L2.

We should also go through the config list to see if that would cover all existing configuration files

**Review by carmocca — CHANGES_REQUESTED** [review]

File: lit_gpt/tokenizer.py
Comment (nkasmanoff) [reply lit_gpt/tokenizer.py:64]: @carmocca  updated, marking as resolved

File: lit_gpt/tokenizer.py
Comment (nkasmanoff) [reply lit_gpt/tokenizer.py:1]: @carmocca let me know what you think of the updated approach, it's all done within the init of the tokenizer class

File: lit_gpt/tokenizer.py
Comment (carmocca) [inline lit_gpt/tokenizer.py:83]:
```suggestion
        if bos or self.use_bos:
```

File: lit_gpt/tokenizer.py
Comment (carmocca) [inline lit_gpt/tokenizer.py:17]: This code path (for sentencepiece tokenizers) will need to load `tokenizer_config.json` to read https://huggingface.co/meta-llama/Llama-2-7b-hf/blob/main/tokenizer_config.json#L2. I don't think `bos_id is not None` is reliable enough

File: lit_gpt/tokenizer.py
Comment (carmocca) [inline lit_gpt/tokenizer.py:37]: "add_bos_token" is only set for sentencepiece tokenizer configs (I think!)

File: lit_gpt/tokenizer.py
Comment (carmocca) [reply lit_gpt/tokenizer.py:17]: Also remember that these things might not be defined. https://huggingface.co/garage-bAInd/Platypus2-70B-instruct/blob/main/tokenizer_config.json is an example of a sentenepiece config that doesn't include this argument.

**Review by carmocca — COMMENTED** [review]
We should add a parametrized test that goes over the different configs and compares the HF tokenizer with ours, as in https://github.com/Lightning-AI/lit-gpt/issues/545#issuecomment-1717661708

File: lit_gpt/tokenizer.py
Comment (nkasmanoff) [reply lit_gpt/tokenizer.py:17]: If not defined I can change it to return false, which should allow the any() to still work. Will push an update to this in a sec

File: tests/test_tokenizer.py
Comment (nkasmanoff) [inline tests/test_tokenizer.py:51]: @carmocca thoughts for this? 

https://github.com/Lightning-AI/lit-gpt/pull/552#pullrequestreview-1631160390

File: lit_gpt/tokenizer.py
Comment (nkasmanoff) [inline lit_gpt/tokenizer.py:31]: @carmocca wasn't sure if I should also change this to apply `check_if_bos_token_used`. My thinking is that if the special tokens path has a bos_token id, we should assume it is used

File: tests/test_tokenizer.py
Comment (carmocca) [reply tests/test_tokenizer.py:51]: I'm adding it in https://github.com/Lightning-AI/lit-gpt/pull/565. You can rebase your on that and resolve the TODO

File: lit_gpt/tokenizer.py
Comment (carmocca) [reply lit_gpt/tokenizer.py:31]: I would call it, the bos_token_id might be defined even if it's not used.

**Review by carmocca — COMMENTED** [review]
I incorporated the test for all tokenizer configs and removed the `xfail` marker. You can now use it to guide you and fix the remaining failures

File: lit_gpt/tokenizer.py
Comment (nkasmanoff) [reply lit_gpt/tokenizer.py]: Updated this to also use the same check, marking this as resolved.

**Comment by nkasmanoff** [comment]
@carmocca Can you re-run the tests please? I did not realize the logs from the failure would go away after a commit and need approval 😅 I'll keep track of them going forward

**Comment by nkasmanoff** [comment]
@carmocca 

Solved most test failure cases, but there are some problems with HF’s format that’s leaving this with some failed tests.
Left some TODOs which explain the reasoning. 

There is one rather significant issue though, which I can share below

For non-sentencepiece tokenizers, i.e. not Llamas, the tokenizer class relies on `generation_config.json` to get the BOS and EOS tokens. 

What I’m noticing then is the BOS token may be set to `1`,  but are `None` for HF. I feel okay about the solution I proposed here which is since `use_bos` checks whether or not to use the BOS token, so it shouldn’t matter what it is set to. 

This leads to a different fail, for falcon models (https://huggingface.co/tiiuae/falcon-7b/blob/main/generation_config.json#L4) where the EOS token is defined differently in their config than it is in the tokenizer file. 
https://github.com/Lightning-AI/lit-gpt/actions/runs/6227679870/job/16902841614#step:5:618


I’d be curious what affect this has had on folks fine-tuning falcon models already, they may have been using the wrong EOS token.

File: lit_gpt/tokenizer.py
Comment (nkasmanoff) [inline lit_gpt/tokenizer.py:74]: Nice!

File: tests/test_tokenizer.py
Comment (carmocca) [inline tests/test_tokenizer.py:66]: @nkasmanoff I added these special checks. As you observed, seems like the HF tokenizer is a bit broken for falcon and stablecode :sweat_smile:

**Review by carmocca — APPROVED** [review]
I ran all tests and they all seems to pass

File: tests/test_tokenizer.py
Comment (nkasmanoff) [inline tests/test_tokenizer.py:54]: Also nice! FYI I submitted a PR on the falcon config files on huggingface, which should fix the issue upstream as well (though it is not clear when that changes if ever) If it does indeed get fixed there, can adjust it here as well.

Comment (carmocca) [reply tests/test_tokenizer.py:54]: What's the link? I'd like to see it

Comment (nkasmanoff) [reply tests/test_tokenizer.py:54]: This, https://huggingface.co/tiiuae/falcon-rw-7b/discussions/4/files, copy and pasted for all falcon variants aside from 180B. 

Looks like the problem was already fixed there: 
https://huggingface.co/tiiuae/falcon-180B-chat/blob/main/generation_config.json

Comment (carmocca) [reply tests/test_tokenizer.py:54]: You seem to have submitted it to a different model. We use https://huggingface.co/tiiuae/falcon-7b (not `falcon-rw-7b`)

**Pull request merged**
